README.md
# 🐶 Pupcloud v0.8.0

Put your files on the browser!

_[Documentation](https://germ.gitbook.io/pupcloud/) - [Demo](https://pupcloud.vercel.app/) - [Roadmap](https://github.com/proofrock/pupcloud/wiki/Roadmap)_

Point pupcloud to a local folder and instantly get its contents exposed as a web application. Browse, manage, share
files... in a truly portable way!

Pupcloud doesn't need a database, [a config file](https://github.com/proofrock/pupcloud/discussions/26), nor it leaves
.dotfiles on the filesystem; it is packaged as a single executable file (written in Go+Svelte+Typescript), so you'll
just need to [download](https://germ.gitbook.io/pupcloud/guides/installation-and-building)
and [run](https://germ.gitbook.io/pupcloud/guides/running-pupcloud) it.

It is an exercise in minimalism: it aims to be as simple as possible, while retaining all the features you would like it
to have. Just run it! It has everything that fits in a pup-sized cloud ;-)

If you are curious, go see the [Demo](https://pupcloud.vercel.app/)!

## Features

- [Web interface](https://germ.gitbook.io/pupcloud/guides/basic-usage) that scales well on mobile;
- Nothing is saved on disk (no database, no .dotfiles...);
- [Authentication](https://germ.gitbook.io/pupcloud/guides/authentication) (also see [below](#auth));
- [Read-only mode](https://germ.gitbook.io/pupcloud/guides/running-pupcloud) for avoiding fs writes;
- [File previews](https://germ.gitbook.io/pupcloud/guides/basic-usage#preview-screen);
- (Revokable) [folder sharing](https://germ.gitbook.io/pupcloud/guides/sharing-a-folder), on a separate URL (also
  see [below](#sharing)):
- "[Branding](https://germ.gitbook.io/pupcloud/guides/running-pupcloud)" (you can specify the title of the app screen).

It doesn't include HTTPS, as this can be done easily (and much more securely) with
a [reverse proxy](https://germ.gitbook.io/pupcloud/guides/reverse-proxy).

See the [ROADMAP](ROADMAP.md) file for a glimpse of what's in store!

## Basic usage

Pupcloud is distributed as a single executable file. Download and unpack the proper file for your OS/arch.

Once done, just execute it with the directory to serve as an argument:

```bash
pupcloud -r /my/dir
```

Then, open `http://localhost:17178` with a browser. As simple as that!

Run `pupcloud --help` to see the other [config options](https://germ.gitbook.io/pupcloud/guides/running-pupcloud). You
can also:

- enable "write" operations (delete/cut/paste/upload...; `-E` or `--allow-edits`);
- setup [authentication](#auth) (`-P`/`-H`);
- setup [folder sharing](#sharing) (`--share-profile`, `--share-port`, `--share-prefix`)
- specify a title/brand for the window (`--title`);
- use a different port then the default of 17178 (`-p`);
- bind to a network interface (`--bind-to`);
- instruct pupcloud to follow symlinks (`--follow-symlinks`);
- specify a maximum size for upload, in Megabytes (`--max-upload-size`).

By default, it's forbidden to run it as root. Use `--allow-root` if you (really) want to.

### Supported file types for viewing

In the file system view, you can click on a file to open it. Currently pupcloud supports:

- Images, Audio, Video (when supported by the browser)
- PDF documents (for desktop browsers)
- Text-like files (txt, html, sources...)

Detection of file types is done by mime type, and viewing relies on the browser's capabilities.

## <a name="auth"></a>Authentication

You can set a password for accessing pupcloud, by using the `-P` or `-H` parameter on the commandline.

Using `-P`, you can specify the password in plain text. There is a more secure option, `-H`, using which you must
provide the SHA-256 sum of the password you want to use, in (lowercase) hex format.

You can provide the whole hash, or just the first part, of any length you want to keep the commandline short. Of course,
the longer the hash, the safer the system.

```bash
# Password is "ciao", with 128 bit of strength (truncated at 16 bytes)
pupcloud -r /my/dir -P b133a0c0e9bee3be20163d2ad31d6248
```

You can use [this site](https://emn178.github.io/online-tools/sha256.html) to hash the password, it doesn't send the
password on the net (at least at the time I am writing, you may want to check).

**BEWARE**: The password is sent in clear text over the net,
so *[always use a HTTPS-capable reverse proxy](https://germ.gitbook.io/pupcloud/guides/reverse-proxy)* if you plan to
serve over the public internet.

## <a name="sharing"></a>Folder sharing

Sharing a folder is possible. Pupcloud will launch a separate server, on another port, to allow to remap it on a reverse
proxy.

In order to set up a share, one or more *profiles* must be specified. Each profile is in the form *name*:*secret*, where
the secret must not be shared with the recipient, and it's used to protect the confidentiality of the share.

In the Web interface, the sharing URL can be obtained using the "share" button, and specifying:

- the profile;
- a password (optional);
- an expiry date (optional);
- if the share must be read-only.

The share can be "revoked" by relaunching pupcloud without a particular profile, or changing the secret of a profile.
All the links tied to that profile will be invalidated. Also, if the main app is launched as read-only, all its share
links will be read-only.

Relevant CLI parameters are:

- `--share-profile`: a share profile, in the form *name*:*secret* (e.g. `Family:abc0123`). Can be repeated for more
  profiles.
- `--share-port`: the port for the share server; by default `17179`;
- `--share-prefix`: useful when using a reverse proxy, it's the base URL of the share link. By
  default, `http://localhost:17179`.

Sharing is enabled if at least one profile is defined.

## Docker

Official docker image available. See [DockerHub's page](https://hub.docker.com/r/germanorizzo/pupcloud) for
instructions.

## Known bugs

- In rare cases, MIME type detection is wrong. It relies on Go builtin functions, so it needs to be investigated more.
- On mobile, some glitches may be present. Please report them in the issue tracker!
- In general, pupcloud is not (yet) mature and well
  audited. [Take appropriate measures](https://app.gitbook.com/s/BIkxAX0ktzzPPM6PIcMj/security) if you want to publish
  it over the public internet!

## Credits

Kindly supported by [JetBrains for Open Source development](https://jb.gg/OpenSourceSupport).

#### Server

- [fiber](https://gofiber.io/) [MIT]
- otiai10's [copy](https://github.com/otiai10/copy) [MIT]
- spf13's [pflag](https://github.com/spf13/pflag) [BSD 3-Clause]

#### Web UI

- [axentix](https://useaxentix.com/) [MIT]
- [hammer.js](https://github.com/hammerjs/hammer.js) [MIT]
- [Material Design Icons](https://materialdesignicons.com/) [Pictogrammers Free]
- [svelte](https://svelte.dev/) [MIT]
- [SweetAlert2](https://github.com/sweetalert2/sweetalert2) [MIT]
- yeyushengfan258's
  [Win10Sur-icon-theme](https://github.com/yeyushengfan258/Win10Sur-icon-theme) [GPLv3]

#### Languages

- [Go](https://go.dev)
- [Typescript](https://www.typescriptlang.org)


Changelog.md
# Changelog

## Main features history

| Feature                                    | Version |
|--------------------------------------------|---------|
| Basic navigation                           | v0.1.x  |
| File preview/gallery                       | v0.1.x  |
| Authentication                             | v0.2.x  |
| Write operations (delete, copy, rename...) | v0.3.x  |
| Read-only mode                             | v0.3.x  |
| File upload                                | v0.4.x  |
| Folder sharing                             | v0.5.x  |
| Follow symlinks                            | v0.6.x  |

## v0.8.0

- Readonly mode is now the default, `-E` to make it read-write [#16]
- Every CLI parameter can be specified as an env variable [addresses #10]
- Added parameter `--share-profiles` to specify profiles as comma-separated [addresses #14]
- Fixed back button management [addresses #15]
- Copying the sharing link to clipboard wouldn't work if not in a secure context [#19]
- Improvements to better support reverse proxies with subpaths [#18, #22]
- Added path to the browser tab's title, so that it figures in the history list
- Reduced number of columns in grid view for xl screens
- Fix: Only the first sharing profile was selectable [#21]
- Fix: Logout button wouldn't show when using a "normal" password [#23]
- Ditch 7z format for distribution, use .tar.gz & zip instead (according to the OS)
- Updates to libraries
- Binaries for win/arm64 and freebsd/amd64

## v0.7.2

- [#1] Support for browser's back/forward buttons navigation
- [#11] Webservices allows paths with '..'
- Docker cross-compiling for amd64, arm, arm64 and multiplatform manifest

## v0.7.1

- [#2] Can't enter directory made by pupcloud
- [#4] When cd to a dir fails, dir name gets appended to the breadcrumb "forever"
- [#5] Provide systemd example/docs (in proofrock/pupcloud-docs)
- [#6] More functional footer
- [#7] Click on title to go to root
- [#8] Listview's whole row should be clickable
- [#9] Listviews' columns should be clickable to change sorting
- Library upgrades

## v0.7.0

- Plain-text password for auth (`-P`)
  - breaking change: old `-P` is now `-H` (for hashed password)
- Shorter sharing links and better detection of failures (root changes after sharing, ...)
  - breaking change: you'll need to re-generate any sharing link you had from previous versions
- Cross-compilation support (removed CGO dependency)
- Manual mimetype resolution for problematic files

## v0.6.4

- IPV6 support
- "splash footer" disappears after 1,5s

## v0.6.1 - 0.6.3

- fixes
- library updates

## v0.6.0

- Ability to follow symbolic links (`--follow-symlinks`)
- Preview now switches between slideshow, full-screen and full size
- Swipe support in slideshow screen, on mobile
- It's now forbidden to run as root (unless `--allow-root`)
- Logout (if password is required)
- Session invalidation when preconditions fail after login (root deletion, expiration...)
- Go 1.18
- Minor fixes and improvements

## v0.5.0

- Folder sharing
- UserID/GroupID support in docker
- Ability to specify max upload size on the CLI
- Many fixes, refactorings and performance improvements

## v0.4.2

- Fixed z-indexes
- Improved performances when listing a huge number of items
- Upgrade to libs

## v0.4.1

- Fix: paste didn't work
- Reload button

## v0.4.0

- File upload
- CSRF protection

## v0.3.0

- "Write" operations
 - Cut/Copy/Paste
 - Delete
 - Rename
 - Create Folder
- Read-only mode (`--readonly`)
- Support for Windows
- Minor UI fixes

## v0.2.0

- Authentication (`-P`);
- Updates to library (particularly Axentix and Fiber);
- Workaround for dropdown triggers (`...`) being rendered over the dropdown
  itself.

## v0.1.2

- Fixes to HTML layout.

## v0.1.1

- Stable demo site;
- Fixes to Dockerfile.

## v0.1.0

- First version;
- Navigation, preview, download of files.


src/main.go
/*
 * This file is part of PupCloud, Copyright (c) 2022-2078, Germano Rizzo
 *
 * PupCloud is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * PupCloud is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with PupCloud.  If not, see <http://www.gnu.org/licenses/>.
 */
package main

import (
	"embed"
	"fmt"
	"github.com/gofiber/fiber/v2/middleware/filesystem"
	"github.com/proofrock/pupcloud/commons"
	"github.com/proofrock/pupcloud/crypgo"
	filess "github.com/proofrock/pupcloud/files"
	"golang.org/x/exp/slices"
	"io/fs"
	"log"
	"net/http"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"sync"
	"time"

	flag "github.com/spf13/pflag"

	"github.com/gofiber/fiber/v2"
	"github.com/gofiber/fiber/v2/middleware/compress"
	"github.com/gofiber/fiber/v2/middleware/csrf"
	"github.com/gofiber/fiber/v2/utils"
)

const Version = "v0.8.0"

const MiB = 1024 * 1024

//go:embed static
var static embed.FS

func errHandler(ctx *fiber.Ctx, err error) error {
	code := fiber.StatusInternalServerError
	msg := err.Error()

	if e, ok := err.(*fiber.Error); ok {
		code = e.Code
	} else {
		if msg[3] == '|' {
			code, _ = strconv.Atoi(msg[:3])
			msg = msg[4:]
		}
	}

	return ctx.Status(code).SendString(msg)
}

type res struct {
	Path  []string      `json:"path"`
	Items []filess.Item `json:"items"`
}

type sharing struct {
	Allowed        bool     `json:"-"`
	ProfileNames   []string `json:"profiles"`
	ProfileSecrets []string `json:"-"`
	Prefix         string   `json:"-"`
}

func main() {
	year := time.Now().Year()
	if year == 2022 {
		fmt.Println(fmt.Sprintf("Pupcloud %s (c) 2022 Germano Rizzo", Version))
	} else {
		fmt.Println(fmt.Sprintf("Pupcloud %s (c) 2022-%d Germano Rizzo", Version, year))
	}

	rootDir := flag.StringP("root", "r", "", "The document root to serve")
	bindTo := flag.String("bind-to", "0.0.0.0", "The address to bind to")
	port := flag.IntP("port", "p", 17178, "The port to run on")
	title := flag.String("title", "🐶 Pupcloud", "Title of the window")
	pwd := flag.StringP("password", "P", "", "The main access password, if desired. Use --pwd-hash for a safer alternative")
	pwdHash := flag.StringP("pwd-hash", "H", "", "SHA256 hash of the main access password, if desired")
	readOnly := flag.Bool("readonly", false, "DEPRECATED: no effect, default is read only. Kept for backwards compatibility.")
	allowEdits := flag.BoolP("allow-edits", "E", false, "Allows changes to FS (default: don't)")
	shareProfiles := flag.StringArray("share-profile", []string{}, "Profile for sharing, in the form name:secret, multiple profiles allowed")
	shareProfilesCSV := flag.String("share-profiles", "", "Profiles for sharing, in the form name:secret, multiple profiles comma-separated")
	sharePrefix := flag.String("share-prefix", "", "The base URL of the sharing interface (default: 'http://localhost:' + the port)")
	sharePort := flag.Int("share-port", 17179, "The port of the sharing interface")
	uploadSize := flag.Int("max-upload-size", 32, "The max size of an upload, in MiB")
	allowRoot := flag.Bool("allow-root", false, "Allow launching as root (default: don't)")
	followLinks := flag.Bool("follow-symlinks", false, "Follow symlinks when traversing directories (default: don't)")

	flag.Parse()

	// If env vars are specified, overwrite the flags

	if _rootDir, present := os.LookupEnv("PUP_ROOT"); present {
		rootDir = &_rootDir
	}
	if _bindTo, present := os.LookupEnv("PUP_BIND_TO"); present {
		bindTo = &_bindTo
	}
	if _strPort, present := os.LookupEnv("PUP_PORT"); present {
		_port, err := strconv.Atoi(_strPort)
		if err != nil {
			commons.Abort("ERROR: env var PUP_PORT should be an integer")
		}
		port = &_port
	}
	if _title, present := os.LookupEnv("PUP_TITLE"); present {
		title = &_title
	}
	if _pwd, present := os.LookupEnv("PUP_PASSWORD"); present {
		pwd = &_pwd
	}
	if _pwdHash, present := os.LookupEnv("PUP_PWD_HASH"); present {
		pwdHash = &_pwdHash
	}
	if _allowEdits, present := os.LookupEnv("PUP_ALLOW_EDITS"); present {
		if _allowEdits == "1" {
			ae := true
			allowEdits = &ae
		}
	}
	if _shareProfilesCSV, present := os.LookupEnv("PUP_SHARE_PROFILES"); present {
		shareProfilesCSV = &_shareProfilesCSV
	}
	if _sharePrefix, present := os.LookupEnv("PUP_SHARE_PREFIX"); present {
		sharePrefix = &_sharePrefix
	}
	if _strSharePort, present := os.LookupEnv("PUP_SHARE_PORT"); present {
		_sharePort, err := strconv.Atoi(_strSharePort)
		if err != nil {
			commons.Abort("ERROR: env var PUP_SHARE_PORT should be an integer")
		}
		sharePort = &_sharePort
	}
	if _strUploadSize, present := os.LookupEnv("PUP_MAX_UPLOAD_SIZE"); present {
		_uploadSize, err := strconv.Atoi(_strUploadSize)
		if err != nil {
			commons.Abort("ERROR: env var PUP_MAX_UPLOAD_SIZE should be an integer")
		}
		uploadSize = &_uploadSize
	}
	if _allowRoot, present := os.LookupEnv("PUP_ALLOW_ROOT"); present {
		if _allowRoot == "1" {
			ar := true
			allowRoot = &ar
		}
	}
	if _followLinks, present := os.LookupEnv("PUP_FOLLOW_SYMLINKS"); present {
		if _followLinks == "1" {
			fl := true
			followLinks = &fl
		}
	}

	// let's continue

	if *readOnly {
		fmt.Fprint(os.Stdout, "WARNING: --readonly is deprecated and will be removed")
	}

	if *pwd != "" && *pwdHash != "" {
		commons.Abort("ERROR: cannot specify both a password and a hashed password")
	}

	if os.Geteuid() == 0 && !*allowRoot {
		commons.Abort("ERROR: running as root is forbidden; use --allow-root if you are really sure")
	}

	if *rootDir == "" {
		commons.Abort("ERROR: you must specify a root dir (-r)")
	}

	if !commons.DirExists(*rootDir) {
		commons.Abort("ERROR: root dir must exist")
	}

	var err error
	*rootDir, err = filepath.Abs(*rootDir)
	if err != nil {
		commons.Abort("ERROR: cannot make root dir absolute")
	}

	sharing := sharing{}

	if *sharePrefix != "" {
		if !(strings.HasPrefix(*sharePrefix, "http://") || strings.HasPrefix(*sharePrefix, "https://")) ||
			strings.HasSuffix(*sharePrefix, "/") {
			commons.Abort(
				"ERROR: malformed '--share-prefix': protocol must be http or https, and it must not end with a '/'",
			)
		}
	} else {
		*sharePrefix = fmt.Sprintf("http://localhost:%d", *sharePort)
	}

	if *shareProfilesCSV != "" {
		if len(*shareProfiles) > 0 {
			commons.Abort("ERROR: cannot specify both '--share-profile' and '--share-profiles'")
		}
		_shareProfiles := strings.Split(*shareProfilesCSV, ",")
		shareProfiles = &_shareProfiles
	}

	if len(*shareProfiles) > 0 {
		sharing.Allowed = true
		sharing.Prefix = *sharePrefix
		for i, tok := range *shareProfiles {
			pos := strings.Index(tok, ":")
			if pos < 0 {
				commons.Abort(fmt.Sprintf("ERROR: malformed profile #%d: it must have a ':'", i+1))
			}
			sharing.ProfileNames = append(sharing.ProfileNames, tok[0:pos])
			sharing.ProfileSecrets = append(sharing.ProfileSecrets, tok[pos+1:])
		}
	}

	fmt.Println(fmt.Sprintf(" - Serving dir %s", *rootDir))
	if *allowEdits {
		fmt.Println("   + Read/Write")
	} else {
		fmt.Println("   + Read Only")
	}
	if *pwdHash != "" {
		fmt.Println("   + With hashed password")
	} else if *pwd != "" {
		fmt.Println("   + With password")
	} else {
		fmt.Println("   + Without password")
	}
	fmt.Println("   + With max upload size:", *uploadSize, "MiB")
	if *followLinks {
		fmt.Println("   + Will follow symbolic links")
	} else {
		fmt.Println("   + Will NOT follow symbolic links")
	}

	if sharing.Allowed {
		fmt.Println(" - Sharing enabled")
		fmt.Println("   + With profiles:", strings.Join(sharing.ProfileNames, ", "))
		fmt.Println("   + At", sharing.Prefix)
		go launchSharingApp(*bindTo, *rootDir, *title, *sharePort, *uploadSize, !*allowEdits, *followLinks, &sharing)
		time.Sleep(1 * time.Second)
	}

	launchMainApp(*bindTo, *rootDir, *title, *pwd, *pwdHash, *port, *uploadSize, !*allowEdits, *followLinks, &sharing)
}

// FIXME limit growth
var sessions sync.Map
var authFailureMutex sync.Mutex

func launchMainApp(bindTo, root, title, pwd, pwdHash string, port, uploadSize int, readOnly, followLinks bool, sharing *sharing) {
	app := fiber.New(
		fiber.Config{
			ErrorHandler:          errHandler,
			DisableStartupMessage: true,
			BodyLimit:             uploadSize * MiB,
			Network:               fiber.NetworkTCP,
		},
	)

	app.Use(compress.New())
	// FIXME: it works, but does it do anything?
	app.Use(csrf.New(csrf.Config{
		CookieSameSite: "Strict",
		Expiration:     24 * time.Hour,
		KeyGenerator:   utils.UUIDv4,
	}))

	subFS, _ := fs.Sub(static, "static")
	app.Use("/", filesystem.New(filesystem.Config{
		Root: http.FS(subFS),
	}))

	app.Use(func(c *fiber.Ctx) error {
		if err := doAuth4MainApp(c, root, pwd, pwdHash); err != nil {
			return err
		}

		c.Locals("root", root)
		c.Locals("title", title)
		c.Locals("readOnly", readOnly)
		c.Locals("sharing", sharing)
		c.Locals("maxRequestSize", uploadSize*MiB)
		c.Locals("hasPassword", pwdHash != "" || pwd != "")
		c.Locals("followLinks", followLinks)
		return c.Next()
	})

	app.Get("/features", features)
	app.Get("/ls", ls)
	app.Get("/file", file)
	app.Get("/logout", logout)
	if sharing != nil && sharing.Allowed {
		app.Get("/shareLink", shareLink)
	}
	if !readOnly {
		app.Delete("/fsOps/del", fsDel)
		app.Post("/fsOps/rename", fsRename)
		app.Post("/fsOps/move", fsMove)
		app.Post("/fsOps/copy", fsCopy)
		app.Put("/fsOps/newFolder", fsNewFolder)
		app.Put("/fsOps/upload", fsUpload)
	}

	time.Sleep(1 * time.Second)

	fmt.Println(fmt.Sprintf(" - Server running on port %d", port))
	log.Fatal(app.Listen(fmt.Sprintf("%s:%d", bindTo, port)))
}

func doAuth4MainApp(c *fiber.Ctx, root, pwd, pwdHash string) error {
	if pwdHash == "" && pwd == "" {
		return nil
	}

	pwdHash = strings.ToLower(pwdHash)

	val := c.Cookies("pupcloud-session")
	if val != "" {
		if _, ok := sessions.Load(val); ok {
			return nil
		}

		if !commons.DirExists(root) {
			return fiber.NewError(498, "Folder doesn't exist anymore")
		}
	}

	pwdFromWeb := c.Get("x-pupcloud-pwd")

	// XXX I use 499 because 401 plus a reverse proxy seems to trigger a Basic Authentication
	// prompt in the browser
	if pwdFromWeb == "" {
		return fiber.NewError(499, "Password required")
	}

	auth := false
	if pwdHash != "" {
		auth = strings.HasPrefix(crypgo.Sha256(pwdFromWeb), strings.ToLower(pwdHash))
	} else {
		auth = pwdFromWeb == pwd
	}

	if !auth {
		authFailureMutex.Lock()
		defer authFailureMutex.Unlock()
		time.Sleep(1 * time.Second)
		return fiber.NewError(499, "Wrong password")
	}

	rnd := utils.UUIDv4()
	cookie := new(fiber.Cookie)
	cookie.Name = "pupcloud-session"
	cookie.Value = rnd
	cookie.Expires = time.Now().Add(24 * time.Hour)
	cookie.SessionOnly = true
	cookie.SameSite = "strict"
	c.Cookie(cookie)
	sessions.Store(rnd, true)

	return nil
}

// Stored in the session map, to recover it from the session cookie
type sharInfo struct {
	root        string
	path        string
	hasPassword bool
	readOnly    bool
	expiry      *uint32
}

func launchSharingApp(bindTo, root, title string, port, uploadSize int, globalReadOnly, followLinks bool, sharing *sharing) {
	app := fiber.New(
		fiber.Config{
			ErrorHandler:          errHandler,
			DisableStartupMessage: true,
			BodyLimit:             uploadSize * MiB,
			Network:               fiber.NetworkTCP,
		},
	)

	app.Use(compress.New())
	// FIXME: it works, but does it do anything?
	app.Use(csrf.New(csrf.Config{
		CookieSameSite: "Strict",
		Expiration:     24 * time.Hour,
		KeyGenerator:   utils.UUIDv4,
	}))

	subFS, _ := fs.Sub(static, "static")
	app.Use("/", filesystem.New(filesystem.Config{
		Root: http.FS(subFS),
	}))

	app.Use(func(c *fiber.Ctx) error {
		sharinfo, err := doAuth4SharingApp(c, root, globalReadOnly, sharing)
		if err != nil {
			return err
		}

		if sharinfo.path != "/" && strings.HasSuffix(sharinfo.path, "/") {
			sharinfo.path = sharinfo.path[:len(sharinfo.path)-2]
		}
		c.Locals("root", sharinfo.path)
		c.Locals("title", title)
		c.Locals("readOnly", sharinfo.readOnly)
		c.Locals("maxRequestSize", uploadSize*MiB)
		c.Locals("hasPassword", sharinfo.hasPassword)
		c.Locals("followLinks", followLinks)
		return c.Next()
	})

	app.Get("/features", features)
	app.Get("/ls", ls)
	app.Get("/file", file)
	app.Get("/logout", logoutSharing)
	app.Delete("/fsOps/del", fsDel)
	app.Post("/fsOps/rename", fsRename)
	app.Post("/fsOps/move", fsMove)
	app.Post("/fsOps/copy", fsCopy)
	app.Put("/fsOps/newFolder", fsNewFolder)
	app.Put("/fsOps/upload", fsUpload)

	fmt.Println(fmt.Sprintf(" - Sharing server running on port %d", port))
	log.Fatal(app.Listen(fmt.Sprintf("%s:%d", bindTo, port)))
}

func doAuth4SharingApp(c *fiber.Ctx, root string, globalReadOnly bool, sharing *sharing) (*sharInfo, error) {
	now, _ := strconv.Atoi(time.Now().Format("20060102"))

	val := c.Cookies("pupcloud-sharing-session")
	if val != "" {
		if si, ok := sessions.Load(val); ok {
			sinfo := si.(*sharInfo)
			if sinfo.expiry != nil && uint32(now) > *sinfo.expiry {
				return nil, fiber.NewError(498, "Link expired")
			}

			if !commons.DirExists(sinfo.path) {
				return nil, fiber.NewError(498, "Shared folder doesn't exist anymore")
			}

			if sinfo.root != root {
				return nil, fiber.NewError(498, "Root has changed")
			}

			return si.(*sharInfo), nil
		}
	}

	profile := c.Query("p")
	if profile == "" {
		return nil, fiber.NewError(498, "No profile specified")
	}

	pwd := c.Get("x-pupcloud-pwd")

	// XXX I use 499/498 because 401 plus a reverse proxy seems to trigger a Basic Authentication
	// prompt in the browser

	prfIdx := slices.Index(sharing.ProfileNames, profile)
	if prfIdx < 0 {
		return nil, fiber.NewError(498, "Unknown profile")
	}
	secret := sharing.ProfileSecrets[prfIdx]

	var password string
	if pwd != "" {
		password = pwd + "|" + secret
	} else {
		password = secret
	}

	if c.Query("r") != crypgo.Sha256(root)[:6] {
		return nil, fiber.NewError(498, "Server root dir changed")
	}

	x := c.Query("x")
	if x == "" {
		return nil, fiber.NewError(498, "No sharing details specified")
	}

	partialPath, readOnly, date, err := commons.DecryptSharingURL(profile, password, root, x)
	readOnly = readOnly || globalReadOnly
	if err != nil {
		authFailureMutex.Lock()
		defer authFailureMutex.Unlock()
		time.Sleep(1 * time.Second)
		return nil, fiber.NewError(499, "Wrong password or invalid address")
	}

	sharinfo := sharInfo{
		root,
		filepath.Join(root, partialPath),
		pwd != "",
		readOnly,
		date,
	}

	if date != nil && uint32(now) > *date {
		return nil, fiber.NewError(499, "Link expired")
	}

	if !commons.DirExists(sharinfo.path) {
		return nil, fiber.NewError(499, "Shared folder doesn't exist anymore")
	}

	rnd := utils.UUIDv4()
	cookie := new(fiber.Cookie)
	cookie.Name = "pupcloud-sharing-session"
	cookie.Value = rnd
	cookie.Expires = time.Now().Add(24 * time.Hour)
	cookie.SessionOnly = true
	cookie.SameSite = "strict"
	c.Cookie(cookie)
	sessions.Store(rnd, &sharinfo)

	return &sharinfo, nil
}
